pkg/controller/walle/controller.go

```diff
@@ -0,0 +1,435 @@
+		return err
+	}
+
+	_, exists := wr.nodeInformerMap.Load(key)
+	if !exists {
+		nodeInformer, err := newNodeInformerForKluster(wr.Clients.Satellites, kluster, make(chan struct{}))
+		if err != nil {
+			return err
+		}
+
+		nodeInformer.SharedIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if kluster, err = wr.getKlusterByKey(key); err != nil {
+					wr.logger.Log(err)
+					return
+				}
+				wr.handlersMux.RLock()
+				defer wr.handlersMux.RUnlock()
+				for _, addHandler := range wr.addEventHandlers {
+					addHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if kluster, err = wr.getKlusterByKey(key); err != nil {
+					wr.logger.Log(err)
+					return
+				}
+				wr.handlersMux.RLock()
+				defer wr.handlersMux.RUnlock()
+				for _, updateHandler := range wr.updateEventHandlers {
+					updateHandler(kluster, oldObj.(*api_v1.Node), newObj.(*api_v1.Node))
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if kluster, err = wr.getKlusterByKey(key); err != nil {
+					wr.logger.Log(err)
+					return
+				}
+				wr.handlersMux.RLock()
+				defer wr.handlersMux.RUnlock()
+				for _, deleteHandler := range wr.deleteEventHandlers {
+					deleteHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+		})
+
+		wr.nodeInformerMap.Store(
+			key,
+			nodeInformer,
+		)
+	}
+	return nil
+}
+
+func (wr *WalleController) getKlusterByKey(key string) (*v1.Kluster, error) {
+	o, exists, err := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("kluster %v was not found", key)
+	}
+	return o.(*v1.Kluster), err
+}
+
+func (wr *WalleController) watchNodesForKluster(kluster *v1.Kluster) error {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return err
+	}
+
+	i, ok := wr.nodeInformerMap.Load(key)
+	if !ok {
+		return fmt.Errorf("no node informer created for kluster %s/%s", kluster.GetNamespace(), kluster.GetName())
+	}
+
+	informer := i.(*NodeInformer)
+	cache.WaitForCacheSync(
+		wr.stopCh,
+		informer.HasSynced,
+	)
+
+	go func(informer *NodeInformer) {
+		wr.informerWg.Add(1)
+		defer wr.informerWg.Done()
+		for {
+			select {
+			case <-wr.stopCh:
+				informer.close()
+				return
+			default:
+				informer.run()
+			}
+		}
+	}(informer)
+
+	return nil
+}
+
+func (wr *WalleController) klusterAddFunc(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterUpdateFunc(cur, old interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(cur)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterDeleteFunc(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+	if i, ok := wr.nodeInformerMap.Load(key); ok {
```

Comment:
Doing this here is a little problematic as it circumvents the `workqueue`. The main benefit of the workout is that it provides retries (not really needed in this case) but also serialises the work that is done for a particular key (a Kluster in our case). So in theory while this will be tearing down the Informer there could still be some other event for the kluster in process by the work queue.

Additionally we have to deal with informer teardown in two places: Here and in the `cleanUpInformers` method. It would be good to consolidate it in one place. This would require to fix the workqueue to do reconciliation for deleted klusters as well (see comment above)

Reply:
Removed this here. All handled by `cleanUpInformers`